Leverage std::strong_ordering where appropriate

iidmsa opened the pull request:
#### c8c2a463c97abfc4a5c3e4c27bb58853a2f41184
<pre>
Leverage std::strong_ordering where appropriate
<a href="https://bugs.webkit.org/show_bug.cgi?id=293090">https://bugs.webkit.org/show_bug.cgi?id=293090</a>
<a href="https://rdar.apple.com/problem/151430109">rdar://problem/151430109</a>

Reviewed by Darin Adler.

Provide clearer semantics for comparison operations.

* Source/WTF/wtf/URL.cpp:
(WTF::differingQueryParameters):
* Source/WTF/wtf/text/AtomString.h:
(WTF::codePointCompare):
* Source/WTF/wtf/text/StringImpl.h:
(WTF::codePointCompare):
* Source/WTF/wtf/text/StringView.cpp:
(WTF::codePointCompare):
* Source/WTF/wtf/text/StringView.h:
* Source/WTF/wtf/text/WTFString.cpp:
(WTF::codePointCompare):
* Source/WTF/wtf/text/WTFString.h:
* Source/WebCore/Modules/indexeddb/IDBKey.cpp:
(WebCore::IDBKey::compare const):
* Source/WebCore/Modules/indexeddb/IDBKeyData.cpp:
(WebCore::operator&lt;=&gt;):

Canonical link: <a href="https://commits.webkit.org/295036@main">https://commits.webkit.org/295036@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c32aa618478ef7034b4ef6db9e8504c23241a8d2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/103797 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/23499 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/13819 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/108989 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/54449 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/105837 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/23851 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/32045 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/78859 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/106803 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/18515 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/93638 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59193 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/18327 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/11687 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/53825 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/96472 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/88086 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/11744 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/111377 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/102408 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/30953 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/22819 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/87862 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/31315 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/89839 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/87520 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22298 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/32405 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/10127 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/25310 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/30881 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/36184 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/126041 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/30675 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/34893 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/34010 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/32236 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WTF/wtf/URL.cpp:
```diff
@@ -1283,7 +1283,7 @@ Vector<KeyValuePair<String, String>> differingQueryParameters(const URL& firstUR
     
     auto compare = [] (const KeyValuePair<String, String>& a, const KeyValuePair<String, String>& b) {
-        if (int result = codePointCompare(a.key, b.key))
+        if (auto result = codePointCompare(a.key, b.key); result != std::strong_ordering::equal)
```
`is_neq(result)`

cdumez commented on Source/WTF/wtf/URL.cpp:
```diff
@@ -1300,7 +1300,7 @@ Vector<KeyValuePair<String, String>> differingQueryParameters(const URL& firstUR
     while (indexInFirstQueryParameters < totalFirstQueryParameters && indexInSecondQueryParameters < totalSecondQueryParameters) {
-        int comparison = compare(firstQueryParameters[indexInFirstQueryParameters], secondQueryParameters[indexInSecondQueryParameters]);
+        auto comparison = compare(firstQueryParameters[indexInFirstQueryParameters], secondQueryParameters[indexInSecondQueryParameters]);
         if (comparison < 0) {
```
`is_lt(comparison)`

cdumez replied:
and `is_gt(comparison)` in the next condition.

darinadler commented on Source/WebCore/Modules/indexeddb/IDBKey.cpp:
```diff
@@ -125,11 +125,7 @@ std::weak_ordering IDBKey::compare(const IDBKey& other) const
-        if (!comparison)
-            return std::weak_ordering::equivalent;
-        return comparison < 0 ? std::weak_ordering::less : std::weak_ordering::greater;
+        return codePointCompare(std::get<String>(m_value), std::get<String>(other.m_value));
```
Please also remove the braces now that there is no local variable here.

darinadler commented on Source/WebCore/Modules/indexeddb/IDBKeyData.cpp:
```diff
@@ -285,11 +285,7 @@ std::weak_ordering operator<=>(const IDBKeyData& a, const IDBKeyData& b)
-        if (!comparison)
-            return std::weak_ordering::equivalent;
-        return comparison < 0 ? std::weak_ordering::less : std::weak_ordering::greater;
+        return codePointCompare(std::get<String>(a.m_value), std::get<String>(b.m_value));
```
Please also remove the braces now that there is no local variable here.

iidmsa committed:
Leverage std::strong_ordering where appropriate
https://bugs.webkit.org/show_bug.cgi?id=293090
rdar://problem/151430109

Reviewed by Darin Adler.

Provide clearer semantics for comparison operations.

* Source/WTF/wtf/URL.cpp:
(WTF::differingQueryParameters):
* Source/WTF/wtf/text/AtomString.h:
(WTF::codePointCompare):
* Source/WTF/wtf/text/StringImpl.h:
(WTF::codePointCompare):
* Source/WTF/wtf/text/StringView.cpp:
(WTF::codePointCompare):
* Source/WTF/wtf/text/StringView.h:
* Source/WTF/wtf/text/WTFString.cpp:
(WTF::codePointCompare):
* Source/WTF/wtf/text/WTFString.h:
* Source/WebCore/Modules/indexeddb/IDBKey.cpp:
(WebCore::IDBKey::compare const):
* Source/WebCore/Modules/indexeddb/IDBKeyData.cpp:
(WebCore::operator<=>):

Canonical link: https://commits.webkit.org/295036@main
